besu/src/main/java/org/hyperledger/besu/services/TraceServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpec;
+import org.hyperledger.besu.ethereum.processing.TransactionProcessingResult;
+import org.hyperledger.besu.ethereum.vm.CachingBlockHashLookup;
+import org.hyperledger.besu.evm.tracing.OperationTracer;
+import org.hyperledger.besu.plugin.Unstable;
+import org.hyperledger.besu.plugin.services.TraceService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** The Trace service implementation. */
+@Unstable
+public class TraceServiceImpl implements TraceService {
+  private static final Logger LOG = LoggerFactory.getLogger(TraceServiceImpl.class);
+
+  private final BlockchainQueries blockchainQueries;
+  private final ProtocolSchedule protocolSchedule;
+
+  /**
+   * Instantiates a new TraceServiceImpl service.
+   *
+   * @param blockchainQueries the blockchainQueries
+   * @param protocolSchedule the protocolSchedule
+   */
+  public TraceServiceImpl(
+      final BlockchainQueries blockchainQueries, final ProtocolSchedule protocolSchedule) {
+    this.blockchainQueries = blockchainQueries;
+    this.protocolSchedule = protocolSchedule;
+  }
+
+  /**
+   * Traces block
+   *
+   * @param blockNumber the block number to be traced
+   * @param tracer an instance of OperationTracer
+   */
+  @Override
+  public void traceBlock(final long blockNumber, final Object tracer) {
+    checkArgument(tracer instanceof OperationTracer, "Invalid OperationTracer");
+    final Optional<Block> block = blockchainQueries.getBlockchain().getBlockByNumber(blockNumber);
+    block.ifPresent(value -> trace(value, (OperationTracer) tracer));
+  }
+
+  private void trace(final Block block, final OperationTracer tracer) {
+    LOG.debug("Tracing block {}", block.toLogString());
+    final List<TransactionProcessingResult> results = new ArrayList<>();
+    Tracer.processTracing(
+        blockchainQueries,
+        block.getHash(),
+        traceableState -> {
+          final Blockchain blockchain = blockchainQueries.getBlockchain();
+          final ChainUpdater chainUpdater = new ChainUpdater(traceableState);
+          final ProtocolSpec protocolSpec = protocolSchedule.getByBlockHeader(block.getHeader());
+          final MainnetTransactionProcessor transactionProcessor =
+              protocolSpec.getTransactionProcessor();
+          final BlockHeader header = block.getHeader();
+          block
+              .getBody()
+              .getTransactions()
+              .forEach(
+                  transaction -> {
+                    final Optional<BlockHeader> maybeParentHeader =
+                        blockchain.getBlockHeader(header.getParentHash());
+                    final Wei dataGasPrice =
+                        protocolSpec
+                            .getFeeMarket()
+                            .dataPrice(
+                                maybeParentHeader
```

Comment:
Why base the dataPrice on prior block instead of the one being traced?

Reply:
That is how the AbstractBlockCreator.createBlock implements. The other traces, block replay and transaction simulation also base the price on the parent of the block being traced/simulated. I believe that when creating/replaying a block you have to use the latest known info, which would be the parent?